File: Sprint7/O/C++/main.cpp
// https://contest.yandex.ru/contest/25596/problems/O/
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

const long long MOD = 1e9 + 7;

int main() {
    int n, m;
    cin >> n >> m;

    vector<vector<int>> graph(n);
    vector<int> in_degree(n, 0);

    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        u--; v--;
        graph[u].push_back(v);
        in_degree[v]++;
    }

    int a, b;
    cin >> a >> b;
    a--; b--;

    vector<long long> dp(n, 0);
    dp[a] = 1;

    queue<int> q;
    for (int i = 0; i < n; i++) {
        if (in_degree[i] == 0) {
            q.push(i);
        }
    }

    while (!q.empty()) {
        int curr = q.front();
        q.pop();

        for (int neighbor : graph[curr]) {
            dp[neighbor] = (dp[neighbor] + dp[curr]) % MOD;
            in_degree[neighbor]--;
            if (in_degree[neighbor] == 0) {
                q.push(neighbor);
            }
        }
    }

    cout << dp[b] << endl;

    return 0;
}

File: Sprint7/O/Rust/src/main.rs
// https://contest.yandex.ru/contest/25596/problems/O/

use std::collections::VecDeque;

fn main() {
    let mut lines = std::io::stdin().lines();
    let line = lines.next().unwrap().unwrap();
    let mut n_m = line.split_whitespace();
    let n = n_m.next().unwrap().parse::<usize>().unwrap();
    let m = n_m.next().unwrap().parse::<usize>().unwrap();

    let mut graph = vec![vec![]; n];
    let mut in_degree = vec![0; n];

    for _ in 0..m {
        let line = lines.next().unwrap().unwrap();
        let mut uv = line.split_whitespace();
        let u = uv.next().unwrap().parse::<usize>().unwrap() - 1;
        let v = uv.next().unwrap().parse::<usize>().unwrap() - 1;
        graph[u].push(v);
        in_degree[v] += 1usize;
    }

    let line = lines.next().unwrap().unwrap();
    let mut ab = line.split_whitespace();
    let a = ab.next().unwrap().parse::<usize>().unwrap() - 1;
    let b = ab.next().unwrap().parse::<usize>().unwrap() - 1;

    let mut dp = vec![0usize; n];
    dp[a] = 1;

    let mut q: VecDeque<usize> = VecDeque::new();
    for (i, d) in in_degree.iter().enumerate() {
        if *d == 0 {
            q.push_back(i);
        }
    }

    while let Some(curr) = q.pop_front() {
        for &neighbor in &graph[curr] {
            dp[neighbor] = (dp[neighbor] + dp[curr]) % usize::MAX;
            in_degree[neighbor] -= 1;
            if in_degree[neighbor] == 0 {
                q.push_back(neighbor);
            }
        }
    }

    println!("{}", dp[b]);
}
